### ISSUE TITLE
Spurious error regarding runtime type values

### ISSUE THREAD

**Issue by gwk** [python/mypy#3066]
mypy eb1eb0d9 errors on the following:
```
def f():
  t = str
  if True:
    t = int
```

test.py:4:4: error: Incompatible types in assignment (expression has type "int", variable has type "str")

This code is valid; the proper type for `x` is `type`. I would hope that the inferencer can figure this out; after all, `type(str) == type`. At the very least the error message is definitely incorrect. It appears to be confusing the type of the right-hand-side with the value.

Adding the annotation `t: type = str` fixes the warning, further evidence that the message is incorrect.

(For the curious, the code is a minimization of a real-world function in which we are choosing and returning a type to be used later as a constructor function.)

**Comment by ilevkivskyi** [comment]
This may be a duplicate of #3050

**Comment by gvanrossum** [comment]
No it smells like a bug beyond the display of types.

**Comment by gwk** [comment]
I think you are both right, in that the first example of #3050 is another example of mypy mixing up the type with an instance of the type.

```
class A: pass
class B: pass
x: A = B
# error: Incompatible types in assignment (expression has type "B", variable has type "A")
```
In the absence of fancier metatypes or anything, this should at least read 'expression has type "type", variable has type "A"'.

**Comment by gvanrossum** [comment]
Yeah, there are two things going on -- the error message is confusing (and that's a duplicate of #3050), and the type initially inferred for `t` is `Type[int]` rather than `type` (which is equivalent to `Type[Any]`). But that is arguably correct, and can be fixed easily by adding `# type: type` to the initial assignment. It's not so different from
```python
a = 1
if bool():
    a = ''  # E: Incompatible types in assignment (expression has type "str", variable has type "int")
```
which is also arguably correct code but rejected by mypy and can by fixed by adding `# type: Any`.

So I'm going to close this because I think we're not going to change mypy's behavior regarding whether this is an error or not, and we already have #3050 for the confusing error.
